File: lib/src/ztgexc.dart
// Copyright (c) 2024 Guilherme Lepsch. All rights reserved. Use of this
// source code is governed by a BSD-style license that can be found in the
// [LICENSE file](https://github.com/lepsch/dart_lapack/blob/main/LICENSE).

import 'dart:math';

import 'package:dart_lapack/src/box.dart';
import 'package:dart_lapack/src/complex.dart';
import 'package:dart_lapack/src/matrix.dart';
import 'package:dart_lapack/src/xerbla.dart';
import 'package:dart_lapack/src/ztgex2.dart';

void ztgexc(
  final bool WANTQ,
  final bool WANTZ,
  final int N,
  final Matrix<Complex> A_,
  final int LDA,
  final Matrix<Complex> B_,
  final int LDB,
  final Matrix<Complex> Q_,
  final int LDQ,
  final Matrix<Complex> Z_,
  final int LDZ,
  final int IFST,
  final Box<int> ILST,
  final Box<int> INFO,
) {
  final A = A_.having(ld: LDA);
  final B = B_.having(ld: LDB);
  final Q = Q_.having(ld: LDQ);
  final Z = Z_.having(ld: LDZ);
  int HERE = 0;

  // Decode and test input arguments.
  INFO.value = 0;
  if (N < 0) {
    INFO.value = -3;
  } else if (LDA < max(1, N)) {
    INFO.value = -5;
  } else if (LDB < max(1, N)) {
    INFO.value = -7;
  } else if (LDQ < 1 || WANTQ && (LDQ < max(1, N))) {
    INFO.value = -9;
  } else if (LDZ < 1 || WANTZ && (LDZ < max(1, N))) {
    INFO.value = -11;
  } else if (IFST < 1 || IFST > N) {
    INFO.value = -12;
  } else if (ILST.value < 1 || ILST.value > N) {
    INFO.value = -13;
  }
  if (INFO.value != 0) {
    xerbla('ZTGEXC', -INFO.value);
    return;
  }

  // Quick return if possible

  if (N <= 1) return;
  if (IFST == ILST.value) return;

  if (IFST < ILST.value) {
    HERE = IFST;

    do {
      // Swap with next one below

      ztgex2(WANTQ, WANTZ, N, A, LDA, B, LDB, Q, LDQ, Z, LDZ, HERE, INFO);
      if (INFO.value != 0) {
        ILST.value = HERE;
        return;
      }
      HERE++;
    } while (HERE < ILST.value);
    HERE--;
  } else {
    HERE = IFST - 1;

    do {
      // Swap with next one above

      ztgex2(WANTQ, WANTZ, N, A, LDA, B, LDB, Q, LDQ, Z, LDZ, HERE, INFO);
      if (INFO.value != 0) {
        ILST.value = HERE;
        return;
      }
      HERE--;
    } while (HERE >= ILST.value);
    HERE++;
  }
  ILST.value = HERE;
}

File: SRC/ztgexc.f
      SUBROUTINE ZTGEXC( WANTQ, WANTZ, N, A, LDA, B, LDB, Q, LDQ, Z, LDZ, IFST, ILST, INFO );

*  -- LAPACK computational routine --
*  -- LAPACK is a software package provided by Univ. of Tennessee,    --
*  -- Univ. of California Berkeley, Univ. of Colorado Denver and NAG Ltd..--

      // .. Scalar Arguments ..
      bool               WANTQ, WANTZ;
      int                IFST, ILST, INFO, LDA, LDB, LDQ, LDZ, N;
      // ..
      // .. Array Arguments ..
      COMPLEX*16         A( LDA, * ), B( LDB, * ), Q( LDQ, * ), Z( LDZ, * );
      // ..

*  =====================================================================

      // .. Local Scalars ..
      int                HERE;
      // ..
      // .. External Subroutines ..
      // EXTERNAL XERBLA, ZTGEX2
      // ..
      // .. Intrinsic Functions ..
      // INTRINSIC MAX
      // ..
      // .. Executable Statements ..

      // Decode and test input arguments.
      INFO = 0;
      if ( N < 0 ) {
         INFO = -3;
      } else if ( LDA < MAX( 1, N ) ) {
         INFO = -5;
      } else if ( LDB < MAX( 1, N ) ) {
         INFO = -7;
      } else if ( LDQ < 1 || WANTQ && ( LDQ < MAX( 1, N ) ) ) {
         INFO = -9;
      } else if ( LDZ < 1 || WANTZ && ( LDZ < MAX( 1, N ) ) ) {
         INFO = -11;
      } else if ( IFST < 1 || IFST > N ) {
         INFO = -12;
      } else if ( ILST < 1 || ILST > N ) {
         INFO = -13;
      }
      if ( INFO != 0 ) {
         xerbla('ZTGEXC', -INFO );
         return;
      }

      // Quick return if possible

      if (N <= 1) RETURN       IF( IFST == ILST ) RETURN;

      if ( IFST < ILST ) {

         HERE = IFST;

         } // 10

         // Swap with next one below

         ztgex2(WANTQ, WANTZ, N, A, LDA, B, LDB, Q, LDQ, Z, LDZ, HERE, INFO );
         if ( INFO != 0 ) {
            ILST = HERE;
            return;
         }
         HERE = HERE + 1;
         if (HERE < ILST) GO TO 10;
         HERE = HERE - 1;
      } else {
         HERE = IFST - 1;

         } // 20

         // Swap with next one above

         ztgex2(WANTQ, WANTZ, N, A, LDA, B, LDB, Q, LDQ, Z, LDZ, HERE, INFO );
         if ( INFO != 0 ) {
            ILST = HERE;
            return;
         }
         HERE = HERE - 1;
         if (HERE >= ILST) GO TO 20;
         HERE = HERE + 1;
      }
      ILST = HERE;
      return;

      // End of ZTGEXC

      }
